envbuilder.go

```diff
@@ -1627,55 +1627,83 @@ func parseMagicImageFile(fs billy.Filesystem, path string, v any) error {
 	return nil
 }
 
-func initDockerConfigJSON(logf log.Func, workingDir workingdir.WorkingDir, dockerConfigBase64 string) (func() error, error) {
-	var cleanupOnce sync.Once
-	noop := func() error { return nil }
+func initDockerConfigOverride(logf log.Func, workingDir workingdir.WorkingDir, dockerConfigBase64 string) (func() error, error) {
+	var (
+		oldDockerConfig = os.Getenv("DOCKER_CONFIG")
+		newDockerConfig = workingDir.Path()
+		cfgPath         = workingDir.Join("config.json")
+		restoreEnv      = func() error { return nil } // noop.
+	)
+	if dockerConfigBase64 != "" || oldDockerConfig == "" {
+		err := os.Setenv("DOCKER_CONFIG", newDockerConfig)
+		if err != nil {
+			logf(log.LevelError, "Failed to set DOCKER_CONFIG: %s", err)
+			return nil, fmt.Errorf("set DOCKER_CONFIG: %w", err)
+		}
+		logf(log.LevelInfo, "Set DOCKER_CONFIG to %s", newDockerConfig)
+
+		restoreEnv = func() error {
+			// Restore the old DOCKER_CONFIG value.
+			if oldDockerConfig == "" {
+				err := os.Unsetenv("DOCKER_CONFIG")
+				if err != nil {
+					err = fmt.Errorf("unset DOCKER_CONFIG: %w", err)
+				}
```

Comment:
suggestion: could this also be simplified?

```
return fmt.Errorf(...)
```

Reply:
Surely! Had a log before, forgot to simplify.